### PR TITLE
fix receiver on calls of imported and exported functions

### PR DESCRIPTION
**Pull request by ajafff**
fixes: #35420

@rbuckton for review

### CODE REVIEW

**Comment by ljharb** [comment]
I see a lot of fixture changes but none of the test cases from #35420 added - specifically, a `this`-sensitive exported function being both called bare, and also `.call`ed with an explicit receiver.

**Comment by ajafff** [comment]
@ljharb the vast majority of the tests only verifies the correct transpilation result by comparing with the baselines. You can see from the changed baselines that imported functions are now called indirectly, which fixes the issue with the receiver.
IMO it's not really necessary to verify the runtime behavior with a test that evaluates the transpiled code. However, if you or someone else has a strong opinion on that, I'll add a test that does it.

**Comment by ljharb** [comment]
It's certainly not up to me, but the important thing is the correct runtime semantics of the code, not the specifics of what gets output, so it seems like those tests would be the important thing :-)

**Comment by ljharb** [comment]
@sandersn 3.8.1 has come and gone (#35420 is listed as in that milestone), and this important bugfix hasn't gone out yet. any update?

**Comment by sandersn** [comment]
@rbuckton This is now scheduled for 4.0. Can you review this fix during the beta?

File: src/compiler/transformers/module/module.ts [hunk 1]
```diff
@@ -1662,6 +1668,28 @@ namespace ts {
             return node;
         }
 
+        function substituteCallExpression(node: CallExpression) {
+            if (!isIdentifier(node.expression)) {
+                return node;
+            }
+            const newExpression = substituteExpressionIdentifier(node.expression);
```
Comment (rbuckton) [inline src/compiler/transformers/module/module.ts:1675]: The substitution for the identifier should be invoked by the printer, you shouldn't have to call it directly.

**Comment by ljharb** [comment]
Ping again - i know this needs a rebase, but it’s a pretty critical bug fix.

**Comment by rbuckton** [comment]
I've rebased this PR and made a few small updates. We may also want to consider doing the same thing for namespace exports:

```ts
namespace X {
  export function foo(this: unknown) {
    console.log(this === X);
  }
}
namespace X {
  foo(); // compiles to `X.foo()`, so would print `true`
}
```

@DanielRosenwasser we should discuss this emit change at the next design meeting.

**Comment by rbuckton** [comment]
I've updated this to support namespace exports as well. as its essentially a bug there too and I'd like to see how this change effects compiler performance (because we call a lot of functions on the `ts` namespace).

**Comment by rbuckton** [comment]
@typescript-bot perf test

**Comment by rbuckton** [comment]
@typescript-bot pack this

**Comment by rbuckton** [comment]
Per discussion with @RyanCavanaugh today, I'm going to revert the change for namespaces for the time being.

**Comment by rbuckton** [comment]
Hmm. build is failing for a unit test that doesn't exist anymore...

**Comment by rbuckton** [comment]
It looks like we will need to revert this change until TS 4.4 while we investigate the significant performance regression this causes during emit. In the worst case, we may need to land this behind a flag to opt-in to the change, though ideally I can find a mitigation for the performance regression so that we can land this without a flag.

**Comment by ljharb** [comment]
Reverting this seems like it will restore the active breakage on a number of my shims like Promise.allSettled, where the module.exports function explicitly relies on having an undefined receiver.

Why does a performance regression matter when compared to runtime breakage?

**Comment by ljharb** [comment]
@rbuckton ping, TS is severely broken without this fix; any update on restoring it?

**Comment by DanielRosenwasser** [comment]
We re-re-discussed this at https://github.com/microsoft/TypeScript/issues/44555.

We've always tried to balance reasonable/canonical-looking emit and general UX with exact semantics of the spec. We're not ready to sacrifice that *and also* impose an emit overhead on everyone, especially when performance is something we're really working to improve. If we can find some optimizations that make this "free-ish", we'd feel better about this feature; otherwise, we'll need to put it behind a flag (like `--downlevelIteration`, `--useDefinedForClassFields`, etc.).

> Why does a performance regression matter when compared to runtime breakage?

The bigger risk here is actually imposing runtime breakage on existing users who accidentally came to rely on the existing behavior. Given that, the safer thing might actually be to put it behind a flag.

**Comment by RyanCavanaugh** [comment]
>  so you’re worried about users who expect that, defying the way it works everywhere else in the JS exosystem

Yes. Improved spec compliance has caused problems before, e.g. when we made exports immutable/nonenumerable (https://github.com/jasmine/jasmine/issues/1817, https://github.com/microsoft/TypeScript/issues/38568).

**Comment by ljharb** [comment]
Alright - can it at least be landed behind a flag, so i can tell users how to unbreak their applications?

**Comment by RyanCavanaugh** [comment]
We're definitely going to figure something out, just trying to work out what exactly that is

**Comment by rbuckton** [comment]
We were able to re-land this in #44624 in a way that does not significantly impact emit performance.

**Comment by ljharb** [comment]
That's amazing, great to hear!

**Pull request merged**